presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java

```diff
@@ -251,8 +286,57 @@ public void testCastFromTimestamp()
         assertFunction("CAST(from_unixtime(1) AS JSON)", JSON, "\"" + sqlTimestamp(1000).toString() + "\"");
     }
 
+    @Test
+    public void testCastWithJsonParse()
+    {
+        // array
+        assertCastWithJsonParse("[\"hello\" , \t \"world\"\n   ]", "ARRAY<VARCHAR>", new ArrayType(VARCHAR), ImmutableList.of("hello", "world"));
+        assertCastWithJsonParse("[1 , 2   ]", "ARRAY<BIGINT>", new ArrayType(BIGINT), ImmutableList.of(1L, 2L));
+        assertCastWithJsonParse("[[1,1], [2,2]]", "ARRAY<ARRAY<INTEGER>>", new ArrayType(new ArrayType(INTEGER)), ImmutableList.of(ImmutableList.of(1, 1), ImmutableList.of(2, 2)));
+        assertCastWithJsonParse("[true, false]", "ARRAY<BOOLEAN>", new ArrayType(BOOLEAN), ImmutableList.of(true, false));
+        assertCastWithJsonParse("[null, null ]", "ARRAY<INTEGER>", new ArrayType(INTEGER), Arrays.asList(null, null));
+        assertCastWithJsonParse("[{\"\":1}]", "ARRAY<MAP<VARCHAR,INTEGER>>", new ArrayType(mapType(VARCHAR, INTEGER)), ImmutableList.of(ImmutableMap.of("", 1)));
+        assertInvalidCastWithJsonParse("[null, null ]", "ARRAY<UNKNOWN>", INVALID_CAST_ARGUMENT);
+        assertInvalidCastWithJsonParse("[true, false]", "ARRAY<INTEGER>", INVALID_CAST_ARGUMENT);
+        assertInvalidCastWithJsonParse("[1, false]", "ARRAY<INTEGER>", INVALID_CAST_ARGUMENT);
+
+        // map
+        assertCastWithJsonParse("{\"a\"\n:1,  \"b\":\t2}", "MAP<VARCHAR,INTEGER>", mapType(VARCHAR, INTEGER), ImmutableMap.of("a", 1, "b", 2));
+        assertCastWithJsonParse("{\"1\" : 2, \"3\":4   }", "MAP<BIGINT,INTEGER>", mapType(BIGINT, INTEGER), ImmutableMap.of(1L, 2, 3L, 4));
+        assertCastWithJsonParse("{\"1\": [2,2]}", "MAP<INTEGER,ARRAY<INTEGER>>", mapType(INTEGER, new ArrayType(INTEGER)), ImmutableMap.of(1, ImmutableList.of(2, 2)));
+        assertCastWithJsonParse("{\"true\": false}", "MAP<BOOLEAN,BOOLEAN>", mapType(BOOLEAN, BOOLEAN), ImmutableMap.of(true, false));
+        Map<String, Integer> nullMap = new HashMap<>();
+        nullMap.put("1", null);
+        assertCastWithJsonParse("{\"1\": null }", "MAP<VARCHAR,INTEGER>", mapType(VARCHAR, INTEGER), nullMap);
+        assertInvalidCastWithJsonParse("{\"[1, 1]\":[2, 2]}", "MAP<ARRAY<INTEGER>,ARRAY<INTEGER>>", INVALID_CAST_ARGUMENT);
+        assertInvalidCastWithJsonParse("{\"true\": false, \"false\":1}", "MAP<BOOLEAN,BOOLEAN>", INVALID_CAST_ARGUMENT);
+
+        // invalid json
+        assertInvalidCastWithJsonParse("{true: false, false:false}", "MAP<BOOLEAN,BOOLEAN>", INVALID_CAST_ARGUMENT);
+        assertInvalidCastWithJsonParse("[1, 2", "ARRAY<INTEGER>", INVALID_CAST_ARGUMENT);
+    }
+
     private static SqlTimestamp sqlTimestamp(long millisUtc)
     {
         return new SqlTimestamp(millisUtc, TEST_SESSION.getTimeZoneKey());
     }
+
+    private void assertCastWithJsonParse(String json, String castSqlType, Type expectedType, Object expected)
+    {
+        // use query runner to avoid constant folding
```

Comment:
I think it's more about "building a query with json strings in VALUES(...)" instead of "use query runner"? 

Constant folding will still happen as long as the query contains `FROM VALUES(...)` (even query runner is used)